### ISSUE TITLE
Cognitive complexity doesn't work for async

### ISSUE THREAD

**Issue by JosuGZ** [rust-lang/rust-clippy#9300]
### Summary

I've been trying the cognitive complexity lint and I have found that on async functions it doesn't work. Here is an example, clippy will complain about `complex`, but not about `complex_async`:

```
#![allow(clippy::all)]
#![deny(clippy::cognitive_complexity)]

pub async fn complex_async() {
    let a = 0;
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
}

pub fn complex() {
    let a = 0;
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
    if a == 0 {}
}
```

### Reproducer

I tried this code:

```rust
<code>
```

I expected to see this happen:

Instead, this happened:

**Comment by guerinoni** [comment]
If I look a good view of code the problem is in `visit_expr`, for some reasons it doesn't count the real complexity but return `1` as complexity for `async` version
